src/ssh/test-sshservice.c

```diff
@@ -0,0 +1,801 @@
+  emit_string (test, NULL, cmd);
+  emit_string (test, "4", "wheee");
+
+  handle_authorize_and_init (test, data);
+
+  while ((sent = mock_transport_pop_channel (test->transport, "4")) == NULL)
+    g_main_context_iteration (NULL, TRUE);
+  cockpit_assert_bytes_eq (sent, "wheee", -1);
+
+  g_object_unref (service);
+  g_free (cmd);
+}
+
+static void
+test_specified_creds_fail (TestCase *test,
+                           gconstpointer data)
+{
+  JsonObject *control = NULL;
+  CockpitSshService *service = NULL;
+
+  service = cockpit_ssh_service_new (COCKPIT_TRANSPORT (test->transport));
+
+  emit_string (test, NULL, "{\"command\": \"init\", \"version\": 1, \"host\": \"localhost\" }");
+  emit_string (test, NULL, "{\"command\": \"open\", \"user\": \"user\","
+                           " \"password\": \"wrong-password\","
+                           " \"host\": \"127.0.0.1\","
+                           " \"channel\": \"4\", \"payload\": \"echo\"}");
+
+  /* Init message */
+  while ((control = mock_transport_pop_control (test->transport)) == NULL)
+    g_main_context_iteration (NULL, TRUE);
+  cockpit_assert_json_eq (control, "{\"command\":\"init\",\"version\":1}");
+  control = NULL;
+
+  /* Should have gotten a failure message, about the credentials */
+  while ((control = mock_transport_pop_control (test->transport)) == NULL)
+    g_main_context_iteration (NULL, TRUE);
+  cockpit_assert_json_eq (control, "{\"command\":\"close\",\"channel\":\"4\",\"problem\":\"authentication-failed\",\"auth-method-results\":{\"password\":\"denied\",\"public-key\":\"denied\",\"gssapi-mic\":\"no-server-support\"}}");
+  control = NULL;
+
+  g_object_unref (service);
+}
+
+static void
+test_host_port (TestCase *test,
+                gconstpointer data)
+{
+  GBytes *sent = NULL;
+  gchar *cmd = NULL;
+  GPid pid;
+  gushort port;
+  CockpitSshService *service = NULL;
+
+  service = cockpit_ssh_service_new (COCKPIT_TRANSPORT (test->transport));
+
+  /* start a new mock sshd on a different port */
+  start_mock_sshd ("auser", "apassword", &pid, &port);
+
+  /* Open a channel with a host that has a port
+   * and a user that doesn't work on the main mock ssh
+   */
+  /* Open a channel with the same user as creds but no password */
+  cmd = g_strdup_printf ("{\"command\": \"open\", \"user\": \"auser\","
+                           " \"password\": \"apassword\","
+                           " \"host\": \"127.0.0.1:%d\","
+                           " \"channel\": \"4\", \"payload\": \"echo\"}", port);
+
+  emit_string (test, NULL, "{\"command\": \"init\", \"version\": 1, \"host\": \"localhost\" }");
+  emit_string (test, NULL, cmd);
+  emit_string (test, "4", "wheee");
+
+  while ((sent = mock_transport_pop_channel (test->transport, "4")) == NULL)
+    g_main_context_iteration (NULL, TRUE);
+  cockpit_assert_bytes_eq (sent, "wheee", -1);
+
+  g_free (cmd);
+  g_object_unref (service);
+
+  stop_mock_sshd (test->mock_sshd);
```

Comment:
This test case also uses `teardown()` which stops the mock sshd again., so I think this can be dropped.